### PR TITLE
feat(testing): add code coverage

### PR DESCRIPTION
**Pull request by jsjoeio**
This solution made possible thanks to @code-asher 

- refactor `jest` setup to play nicely with TS + `lib/vscode`
- add code coverage support
- add code coverage badge in readme

image

### CODE REVIEW

File: ci/dev/test.sh [hunk 1]
```diff
@@ -6,9 +6,11 @@ main() {
 
   cd test/test-plugin
   make -s out/index.js
-  cd "$OLDPWD/test"
-  yarn
-  yarn test "$@"
+  # We must keep jest in a sub-directory. See ../../test/package.json for more
+  # information. We must also run it from the root otherwise coverage will not
+  # include our source files.
+    cd "$OLDPWD"
```
Comment (jsjoeio) [inline ci/dev/test.sh:12]:
```suggestion
  cd "$OLDPWD"
```

File: lib/vscode/build/tsconfig.json [hunk 2]
```diff
@@ -15,12 +15,12 @@
 		"strict": true,
 		"noUnusedLocals": true,
 		"noUnusedParameters": true,
-		"newLine": "lf"
+		"newLine": "lf",
```
Comment (jsjoeio) [inline lib/vscode/build/tsconfig.json:18]: Unnecessary?

File: lib/vscode/build/tsconfig.json [hunk 3]
```diff
@@ -15,12 +15,12 @@
...
 	},
 	"include": [
 		"**/*.ts"
 	],
 	"exclude": [
-		"node_modules/**"
+		"node_modules/**",
```
Comment (jsjoeio) [inline lib/vscode/build/tsconfig.json:24]: Unnecessary?

File: package.json [hunk 4]
```diff
@@ -55,6 +55,7 @@
     "eslint-plugin-import": "^2.18.2",
     "eslint-plugin-prettier": "^3.1.0",
     "leaked-handles": "^5.2.0",
+    "node-fetch": "^2.6.1",
```
Comment (jsjoeio) [inline package.json:58]: I forgot do we need this here or is it only used in tests?

File: package.json [hunk 5]
```diff
@@ -37,7 +37,6 @@
     "@types/http-proxy": "^1.17.4",
     "@types/js-yaml": "^3.12.3",
     "@types/node": "^12.12.7",
-    "@types/node-fetch": "^2.5.7",
```
Comment (jsjoeio) [inline package.json:40]: Moved to `./test/package.json`.

File: ci/dev/test.sh [hunk 6]
```diff
@@ -3,12 +3,24 @@ set -euo pipefail
...
+  if [[ -d ./test/node_modules ]]
+  then
+    echo "Found node_modules directory under /test"
+  else
+    echo "Did not find node_modules directory under /test"
+    echo "...Installing test dependencies."
+    pushd test
+    yarn
+    popd
+  fi
```
Comment (jsjoeio) [inline ci/dev/test.sh:21]: Now that I'm thinking about it...Should we just modify the `postinstall` script to run `yarn` after installing the `lib/vscode` dependencies? 

Pro of that approach:
- no need to do this here

Pro of current approach:
- test.sh doesn't rely on a postinstall script

I could go either way

**Review by code-asher — APPROVED** [review]
Looks good!

**Pull request merged**